aten/src/ATen/core/CachingHostAllocator.h

```diff
@@ -239,30 +277,43 @@ struct CachingHostAllocatorImpl {
       return {nullptr, nullptr};
     }
 
+    auto&& [mempool_id, pool] = get_allocation_pool(get_current_stream());
+
     // If we are using background threads, we can process events in the
     // background.
     if (!pinned_use_background_threads()) {
-      process_events();
+      process_events(pool);
     }
 
     // Round up the allocation to the nearest power of two to improve reuse.
     // These power of two sizes are also used to index into the free list.
     size_t roundSize = c10::llvm::PowerOf2Ceil(size);
 
-    // First, try to allocate from the free list
-    auto* block = get_free_block(roundSize);
+    // First, try to allocate from the free list of the chosen pool
+    auto* block = get_free_block(roundSize, pool);
     if (block) {
+      block->allocating_stream_ = get_current_stream();
+      block->was_allocated_during_stream_capture_ = stream_is_capturing(*block->allocating_stream_);
       return {block->ptr_, reinterpret_cast<void*>(block)};
     }
 
     // Check in the recently freed blocks with pending events to see if we
     // can reuse them. Call get_free_block again after processing events
     if (pinned_use_background_threads()) {
+      process_events_for_specific_size(roundSize, pool);
+      block = get_free_block(roundSize, pool);
+      if (block) {
+        block->allocating_stream_ = get_current_stream();
+        block->was_allocated_during_stream_capture_ = stream_is_capturing(*block->allocating_stream_);
+        return {block->ptr_, reinterpret_cast<void*>(block)};
+      }
+
       // Launch the background thread and process events in a loop.
       static bool background_thread_flag [[maybe_unused]] = [this] {
         getBackgroundThreadPool()->run([&]() {
           while (active_) {
-            process_events();
+            // Background thread conservatively processes default pool events.
+            process_events(default_pool_);
```

Comment:
Should the background thread also process events from all private pools as well?